###*To download the latest version go to the [releases page](https://github.com/gantzgraf/MutationMapper/releases).*

# MutationMapper

MutationMapper is built to help simplify determining the functional consequences of mutations discovered using low-throughput sequencing methods (e.g. Sanger sequencing) and to map CDS coordinates to genomic positions. For details of how to use MutationMapper see the instructions PDF available using the 'help' option in the application or available [here](https://github.com/gantzgraf/MutationMapper/blob/master/src/com/github/mutationmapper/instructions.pdf).


MutationMapper uses Ensemblâ€™s REST API to search genes, sequences and determine functional consequences of mutations. It therefore only works when an internet connection is available.

*Download the latest version for your platform on the [releases page](https://github.com/gantzgraf/MutationMapper/releases).*

###CREDIT

MutationMapper was written by David A. Parry, University of Leeds/University of Edinburgh

If you use this program for a publication, please cite the URL 'https://github.com/gantzgraf/MutationMapper'.

###LICENSE

MutationMapper is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
